src/routers/alpha-router/alpha-router.ts

```diff
@@ -129,41 +130,67 @@ export class AlphaRouter implements IRouter<AlphaRouterConfig>, ISwapToRatio<Alp
         [token0Balance, token1Balance] = [token1Balance, token0Balance]
       }
 
-      const precision = JSBI.BigInt('1' + '0'.repeat(18))
-
-      const sqrtPriceX96 = position.pool.sqrtRatioX96
-      const sqrtPriceX96Lower = TickMath.getSqrtRatioAtTick(position.tickLower)
-      const sqrtPriceX96Upper = TickMath.getSqrtRatioAtTick(position.tickUpper)
-
-      const token0Proportion = SqrtPriceMath.getAmount0Delta(
-        sqrtPriceX96,
-        sqrtPriceX96Upper,
-        precision,
-        true
-      )
-      const token1Proportion = SqrtPriceMath.getAmount1Delta(
-        sqrtPriceX96,
-        sqrtPriceX96Lower,
-        precision,
-        true
+      let optimalRatio = this.calculateOptimalRatio(
+        position,
+        position.pool.sqrtRatioX96
       )
 
-      const amountToSwap = calculateRatioAmountIn(
-        new Fraction(token0Proportion, token1Proportion),
-        position.pool.token0Price,
-        token0Balance,
-        token1Balance,
-      )
+      const zeroForOne = new Fraction(token0Balance.quotient, token1Balance.quotient).greaterThan(optimalRatio)
+      if (!zeroForOne) optimalRatio = optimalRatio.invert()
+
+      const [inputBalance, outputBalance] = zeroForOne
+        ? [token0Balance, token1Balance]
+        : [token1Balance, token0Balance]
+
+      let exchangeRate: Fraction = zeroForOne ? position.pool.token0Price : position.pool.token1Price
+      let swap: SwapRoute<TradeType.EXACT_INPUT> | null = null
+      let ratioAchieved = false
+
+      while (!ratioAchieved) {
+        let amountToSwap = calculateRatioAmountIn(
+          optimalRatio,
+          exchangeRate,
+          inputBalance,
+          outputBalance,
+        )
+
+        swap = await this.routeExactIn(
+          inputBalance.currency,
+          outputBalance.currency,
+          amountToSwap,
+          swapConfig,
+          routingConfig
+        )
+        if (!swap) {
+          throw new Error('did not fetch swap')
+        }
+
+        let inputBalanceUpdated = inputBalance.subtract(swap.trade.inputAmount)
+        let outputBalanceUpdated = outputBalance.add(swap.trade.outputAmount)
+        let newRatio = inputBalanceUpdated.divide(outputBalanceUpdated)
+
+        swap?.route.forEach(route => {
+          route.route.pools.forEach((pool, i) => {
+            if(
+              pool.token0 == position.pool.token0 &&
+              pool.token1 == position.pool.token1 &&
+              pool.fee == position.pool.fee
+            ) {
+              optimalRatio = this.calculateOptimalRatio(
+                position,
+                JSBI.BigInt(route.sqrtPriceX96AfterList[i]!.toString())
+              )
+            }
+          })
+        })
+
+        ratioAchieved =
+          newRatio.equalTo(optimalRatio) ||
+          this.absoluteValue(newRatio.asFraction.divide(optimalRatio).subtract(1)).lessThan(errorTolerance)
+        exchangeRate = swap.trade.outputAmount.divide(swap.trade.inputAmount)
```

Comment:
Originally when we set exchangeRate we used the price of the pool we are adding liquidity for, but here we use the price implied by the swap (which may have been a multi-hop/split route).

Do we even need to update this at all? I thought exchange rate represents the starting price of the pool, so whenever we call `calculateRatioAmountIn` I expected it to be constant?

Reply:
If put the same exchange rate back into the calculation, then we'll swap the same amount again. The iteration is mostly based on the fact that we did not get the amountOut we expected. So we try again at the exchange rate from the last trade, so that we have a better shot of calculating the correct amountOut.

Does that make sense?

Reply:
That makes sense, thanks